modules/cudacodec/include/opencv2/cudacodec.hpp

```diff
@@ -66,181 +66,288 @@ using namespace cuda;  // Stream
 
 ////////////////////////////////// Video Encoding //////////////////////////////////
 
-// Works only under Windows.
-// Supports only H264 video codec and AVI files.
+/** @brief Codecs supported by Video writer, refer to Nvidia's GPU Support Matrix to confirm your GPU supports them.
+*/
+enum class VideoWriterCodec
+{
+    H264 = 0,
+    HEVC = 1,
+    NumCodecs = 2
+};
+
+/** @brief OpenCV color formats.
+*/
+enum COLOR_FORMAT_CV {
+    UNDEFINED = 0,
+    BGR = 1, //!< Default OpenCV color format.
+    RGB = 2,
+    BGRA = 3,
+    RGBA = 4,
+    GRAY = 5
+};
 
-enum SurfaceFormat
+/** @brief Nvidia Video Codec SDK surface formats.
+*/
+enum ENC_BUFFER_FORMAT
 {
-    SF_UYVY = 0,
-    SF_YUY2,
-    SF_YV12,
-    SF_NV12,
-    SF_IYUV,
-    SF_BGR,
-    SF_GRAY = SF_BGR
+    BF_UNDEFINED = 0x00000000, //!< Undefined buffer format.
+    BF_NV12 = 0x00000001, //!< Semi-Planar YUV [Y plane followed by interleaved UV plane].
+    BF_YV12 = 0x00000010, //!< Planar YUV [Y plane followed by V and U planes].
+    BF_IYUV = 0x00000100, //!< Planar YUV [Y plane followed by U and V planes].
+    BF_YUV444 = 0x00001000, //!< Planar YUV [Y plane followed by U and V planes].
+    BF_YUV420_10BIT = 0x00010000, //!< 10 bit Semi-Planar YUV [Y plane followed by interleaved UV plane]. Each pixel of size 2 bytes. Most Significant 10 bits contain pixel data.
+    BF_YUV444_10BIT = 0x00100000, //!< 10 bit Planar YUV444 [Y plane followed by U and V planes]. Each pixel of size 2 bytes. Most Significant 10 bits contain pixel data.
+    BF_ARGB = 0x01000000, //!< 8 bit Packed A8R8G8B8. This is a word-ordered format where a pixel is represented by a 32-bit word with B in the lowest 8 bits, G in the next 8 bits, R in the 8 bits after that and A in the highest 8 bits.
+    BF_ARGB10 = 0x02000000, //!< 10 bit Packed A2R10G10B10. This is a word-ordered format where a pixel is represented by a 32-bit word with B in the lowest 10 bits, G in the next 10 bits, R in the 10 bits after that and A in the highest 2 bits.
+    BF_AYUV = 0x04000000, //!< 8 bit Packed A8Y8U8V8. This is a word-ordered format where a pixel is represented by a 32-bit word with V in the lowest 8 bits, U in the next 8 bits, Y in the 8 bits after that and A in the highest 8 bits.
+    BF_ABGR = 0x10000000, //!< 8 bit Packed A8B8G8R8. This is a word-ordered format where a pixel is represented by a 32-bit word with R in the lowest 8 bits, G in the next 8 bits, B in the 8 bits after that and A in the highest 8 bits.
+    BF_ABGR10 = 0x20000000, //!< 10 bit Packed A2B10G10R10. This is a word-ordered format where a pixel is represented by a 32-bit word with R in the lowest 10 bits, G in the next 10 bits, B in the 10 bits after that and A in the highest 2 bits.
 };
```

Comment:
If I understand correctly, `COLOR_FORMAT_CV` and `ENC_BUFFER_FORMAT` has strong correlation. In case if user define OpenCV style color then particular buffer format should be used. It means, that 2 options are redundant and only OpenCV style format should be presumed as API. Also ARGB10, ABGR10 and other 10 bit  options are not achievable with OpenCV and I do not see a reason to expose them as API.

Reply:
I wanted to add surface types
1) for users who are familiar with Nvidia's Video Coding SDK (maybe unecessary) and
2) because apart from ARGB and BGRA they are not the same format as in OpenCV.

I could add the 8 bit versions to the existing COLOR_FORMAT_CV enum and rename it to COLOR_FORMAT as
enum COLOR_FORMAT {
    UNDEFINED = 0,
    BGR = 1, //!< Default OpenCV color format.
    RGB = 2,
    BGRA = 3,
    RGBA = 4,
    GRAY = 5,

    NV_NV12,
    NV_YV12,
    etc. 
};

which would reduce the number of factory methods while allowing users who have manually created a YV12 `Mat/GpuMat` to use the encoder.

Do you see any problem/issues with this approach?